File: csharp/13/02/Tests/BaseTest.cs
using System;
using NUnit.Framework;
using OpenQA.Selenium;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.IE;
using OpenQA.Selenium.Edge;
using OpenQA.Selenium.Remote;

namespace Tests
{
    [TestFixture]
    class BaseTest
    {
        private static string VendorDirectory = System.IO.Directory.GetParent(
            System.AppContext.BaseDirectory).
            Parent.Parent.Parent.FullName
            + @"/vendor";
        protected IWebDriver Driver;
        public static string BaseUrl;
        private static string Host;
        private static string BrowserName;
        private static string BrowserVersion;
        private static string PlatformName;

        [SetUp]
        protected void SetUp()
        {
            BaseUrl         = System.Environment.GetEnvironmentVariable("BASE_URL") ?? "http://the-internet.herokuapp.com";
            Host            = System.Environment.GetEnvironmentVariable("HOST") ?? "saucelabs";
            BrowserName     = System.Environment.GetEnvironmentVariable("BROWSER_NAME") ?? "ie";
            BrowserVersion  = System.Environment.GetEnvironmentVariable("BROWSER_VERSION") ?? "10.0";
            PlatformName    = System.Environment.GetEnvironmentVariable("PLATFORM_NAME") ?? "Windows 8";
            var sauceUsername = System.Environment.GetEnvironmentVariable("SAUCE_USERNAME");
            var sauceAccessKey = System.Environment.GetEnvironmentVariable("SAUCE_ACCESS_KEY");
            var url = new Uri($"http://{sauceUsername}:{sauceAccessKey}@ondemand.saucelabs.com:80/wd/hub");
            switch (Host.ToLower())
            {
                case "localhost":
                    switch (BrowserName.ToLower())
                    {
                        case "firefox":
                        {
                            var Service = FirefoxDriverService.CreateDefaultService(VendorDirectory);
                            Driver = new FirefoxDriver(Service);
                            break;
                        }
                        case "chrome":
                        {
                            var Service = ChromeDriverService.CreateDefaultService(VendorDirectory);
                            Driver = new ChromeDriver(Service);
                            break;
                        }
                    }
                    break;
                case "saucelabs":
                    switch (BrowserName.ToLower())
                    {
                        case "chrome":
                        {
                            ChromeOptions options = new ChromeOptions();
                            options.PlatformName = PlatformName;
                            options.BrowserVersion = BrowserVersion;
                            Driver = new RemoteWebDriver(url, options.ToCapabilities());
                            break;
                        }
                        case "edge":
                        {
                            EdgeOptions options = new EdgeOptions();
                            options.PlatformName = PlatformName;
                            options.BrowserVersion = BrowserVersion;
                            Driver = new RemoteWebDriver(url, options.ToCapabilities());
                            break;
                        }
                        case "firefox":
                        {
                            FirefoxOptions options = new FirefoxOptions();
                            options.PlatformName = PlatformName;
                            options.BrowserVersion = BrowserVersion;
                            Driver = new RemoteWebDriver(url, options.ToCapabilities());
                            break;
                        }
                        case "ie":
                        {
                            InternetExplorerOptions options = new InternetExplorerOptions();
                            options.PlatformName = PlatformName;
                            options.BrowserVersion = BrowserVersion;
                            Driver = new RemoteWebDriver(url, options.ToCapabilities());
                            break;
                        }
                    }
                    break;
            }
        }
        [TearDown]
        protected void TearDown()
        {
            if (Host.Equals("saucelabs"))
            {
                var testName = TestContext.CurrentContext.Test.Name;
                ((IJavaScriptExecutor)Driver).ExecuteScript("sauce:job-name=" + testName);
            }
            Driver.Quit();
        }
    }
}

File: java/Quickstart/Mod3/Mod3_Final/src/test/java/tests/BaseTest.java
// filename: tests/BaseTest.java
package tests;

import com.saucelabs.saucerest.DataCenter;
import com.saucelabs.saucerest.SauceREST;
import io.github.bonigarcia.wdm.WebDriverManager;
import org.junit.Rule;
import org.junit.rules.ExternalResource;
import org.junit.rules.TestRule;
import org.junit.rules.TestWatcher;
import org.junit.runner.Description;
import org.openqa.selenium.MutableCapabilities;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.edge.EdgeOptions;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.ie.InternetExplorerOptions;
import org.openqa.selenium.remote.BrowserType;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.openqa.selenium.safari.SafariOptions;

import java.net.URL;
import java.util.Date;

import static tests.Config.*;


public class
BaseTest {

    protected WebDriver driver;
    private String testName;
    private String sessionId;
    private SauceREST sauceClient;

    @Rule
    public ExternalResource resource = new ExternalResource() {


        @Override
        protected void before() throws Exception {
            String sauceUrl = "https://ondemand.us-west-1.saucelabs.com/wd/hub";
            MutableCapabilities capabilities;
            switch (browserName) {
                case BrowserType.SAFARI: {
                    capabilities = new SafariOptions();
                    break;
                }
                case BrowserType.FIREFOX: {
                    capabilities = new FirefoxOptions();
                    break;
                }
                case BrowserType.IE: {
                    capabilities = new InternetExplorerOptions();
                    break;
                }
                case BrowserType.EDGE: {
                    capabilities = new EdgeOptions();
                    break;
                }
                default: {
                    capabilities = new ChromeOptions();
                    break;
                }
            }
            capabilities.setCapability("browserVersion", browserVersion);
            capabilities.setCapability("platformName", platformName);

            MutableCapabilities sauceOptions = new MutableCapabilities();
            sauceOptions.setCapability("username", sauceUser);
            sauceOptions.setCapability("accessKey", sauceKey);
            sauceOptions.setCapability("name", testName);

            switch (host) {
                case "saucelabs": {
                    sauceOptions.setCapability("extendedDebugging", "true");
                    sauceOptions.setCapability("capturePerformance", "true");
                    capabilities.setCapability("sauce:options", sauceOptions);

                    driver = new RemoteWebDriver(new URL(sauceUrl), capabilities);
                    sessionId = ((RemoteWebDriver) driver).getSessionId().toString();
                    System.out.printf("Started %s", new Date().toString());
                    System.out.printf("SauceOnDemandSessionID=%s job-name=%s", sessionId, testName);
                    sauceClient = new SauceREST(sauceUser, sauceKey, DataCenter.US);

                    break;
                }
                case "localhost":
                    if ("firefox".equals(browserName)) {
                        WebDriverManager.firefoxdriver().setup();
                        driver = new FirefoxDriver();
                    } else if ("chrome".equals(browserName)) {
                        WebDriverManager.chromedriver().setup();
                        driver = new ChromeDriver();
                    }
                    break;
                case "saucelabs-tunnel": {
                    sauceOptions.setCapability("tunnelIdentifier", sauceTunnel);
                    capabilities.setCapability("sauce:options", sauceOptions);

                    driver = new RemoteWebDriver(new URL(sauceUrl), capabilities);
                    sessionId = ((RemoteWebDriver) driver).getSessionId().toString();
                    System.out.printf("Started %s", new Date().toString());
                    System.out.printf("SauceOnDemandSessionID=%s job-name=%s", sessionId, testName);
                    //these three options copied from Sauce Bindings, need to not reference this.
//                    String sauceReporter = String.format("SauceOnDemandSessionID=%s job-name=%s", this.driver.getSessionId(), this.sauceOptions.sauce().getName());
//                    String sauceTestLink = String.format("Test Job Link: https://app.saucelabs.com/tests/%s", this.driver.getSessionId());
//                    System.out.print(sauceReporter + "\n" + sauceTestLink + "\n");
                    sauceClient = new SauceREST(sauceUser, sauceKey, DataCenter.US);
                    break;
                }
            }
        }

        @Override
        protected void after() {
            driver.quit();

        }

    };
    @Rule
    public TestRule watcher;{
        watcher = new TestWatcher() {
            @Override
            protected void starting(Description description) {
                testName = description.getDisplayName();
            }
            //2 overrides below added
            @Override
            protected void failed(Throwable throwable, Description description) {
                if ("saucelabs".equals(host)) {
                    sauceClient.jobFailed(sessionId);
                    System.out.println(String.format("https://saucelabs.com/tests/%s", sessionId));
                }
            }
            @Override
            protected void succeeded(Description description) {
                if ("saucelabs".equals(host)) {
                    sauceClient.jobPassed(sessionId);
                }
            }
        };
    }
}
